### DISCUSSION TITLE
ItemsControl.ItemsSource ???

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#10867] [General]
Now, in 11P6, we have `ItemsControl.ItemsSource` and `ItemsControl.Items`.

The documentation says this:

> You should use the Avalonia.Controls.ItemsControl.Items property when you're
> assigning a collection of item containers directly, for example adding a collection
> of Avalonia.Controls.ListBoxItems directly to a Avalonia.Controls.ListBox. Add
> the containers to the pre-existing list, do not reassign the Avalonia.Controls.ItemsControl.Items
> property via the setter or with a binding.
> • You should use the Avalonia.Controls.ItemsControl.ItemsSource property when
> you're assigning or binding a collection of models which will be transformed
> by a data template.

OK, I think I understand one is for "adding" and the other is for "assigning".

But I find this extremely confusing and don't understand why. Was it really necessary to add a duplicate property like this?

Could a "copy operation" not be performed in the setter when assigning directly to `ItemsControl.Items`?

**Comment by robloo** [comment]
@kuiperzone The main reasons for this change are two:

 1. Align API with WPF
 2. Performance

See the discussions on it here: 
 * https://github.com/AvaloniaUI/Avalonia/issues/7553 (explains performance reasons)
 * [PR] https://github.com/AvaloniaUI/Avalonia/pull/10590

In your code just change all usages of .Items to .ItemsSource. It is equivalent for this change. A find and replace will do it.

**Reply by kuiperzone** [reply]
Thank you for the reply.

Forgive me, but I find it amusing how you can't please everyone when, in your ref discussion link, you actually wrote:

>Having only a single Items property is intentional in Avalonia. It was originally considered somewhat confusing that there are two properties for this in WPF.

Best wishes
